### PR TITLE
Implement server Stop method

### PR DESCRIPTION
**Pull request by dmitrievanthony**
Currently "Stop" function is not implemented in `GrpcServer`. It leads to the fact, that `GrpcServer` can't be used inside other processes (like in #23022). I've implemented stop method so that server stops and doesn't abort the process during destructor execution.

@mrry, could you please have a look as the author of this functionality?

### CODE REVIEW

File: tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc [hunk 1]
```diff
@@ -398,8 +398,13 @@ Status GrpcServer::Stop() {
...
     case STARTED:
-      return errors::Unimplemented(
-          "Clean shutdown is not currently implemented");
+      server_->Shutdown();
+      master_service_->Shutdown();
+      worker_service_->Shutdown();
+      eager_service_->Shutdown();
+      state_ = STOPPED;
+      LOG(INFO) << "Server stopped (target: " << target() << ")";
+      return Status::OK();
```
Comment (mrry) [inline tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc]: I don't think this will lead to a "clean" shutdown. The various services might have requests in flight, and one would need to ensure that they neither leak memory nor hang forever.

Comment (dmitrievanthony) [reply tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc]: Do I understand correctly that requests are not accepted since `server_->Shutdown()` call? After this shutdown I stops services. Each service continues processing all his enqueued messages and stops after that. Am I right?

Comment (dmitrievanthony) [reply tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc]: Okay, @mrry, do you have any ideas how to make it right? It's really important because without this `stop` functionality we can't properly use server inside other processes.

**Comment by dmitrievanthony** [comment]
I found one more issue, eager service breaks loop immediately after "stop" message. Looks like it's wrong, so I changed behavior to be similar to other services (master and worker).

File: tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc (hunk 1 above)
Comment (mrry) [reply tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc]: No, I haven't worked on the gRPC part of TensorFlow for many years.

Comment (dmitrievanthony) [reply tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc]: I see, @mrry. Anyway, could you please help me a bit with this part? Currently I have the following behavior: when I start and stop server iteratively it fails in ~0.1% cases. It hangs on reading from the queue after shutdown (on `Next()` method).

I've also found out that request enqueueing happens asynchronously and sometimes after shutdown (actually I see that `ENQUEUE_REQUEST` guarded my mutex, but as result of async writing message enqueued after it). Do you know how to fix it? Can we write the message synchronously?

Comment (dmitrievanthony) [reply tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc]: Generally speaking, the requirements are very basic. I need to be able to start and stop gRPC servers inside other process without restarting this process.

**TL;DR** The situation is the following. I'd like to add an ability to delegate TensorFlow cluster manager functionality to Apache Ignite cluster manager system (it's java based platform for distributed computing). To do that I basically need to be able to start/stop TensorFlow server inside Apache Ignite Node Process. First of all, we don't have an ability to start gRPC server from Java, but I'm about to add this functionality in #23022. The only one thing stops me from using TensorFlow gRPC server from Java and implement to whole idea is that `stop()` doesn't work so far.

Comment (dmitrievanthony) [reply tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc]: Speaking about 'lame duck' approach you described, I think I do exactly these things. I stop gRPC server first of all (`server_->Shutdown()`), then stops master, worker and eager services (`master_service_->Shutdown()`). This leads to enqueueing "SHUTDOWN" alarm message into queue and when this message is read from the queue using `cq_->Next()` method I shutdown the queue. Meanwhile I continue reading trailing messages until `cq_Next()` returns `nullptr` (it works because `Thread` destructor requires the loop with `cq_->Next()` to be completed).

Comment (poxvoculi) [reply tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc]: I am not familiar with Apache Ignite.  I looked at a little bit of documentation but don't have time to look into it deeply.  We need to be clear about what 'process' means.  Do you want to kill and restart a gRPC server in the same linux process?  Or does Ignite actually kill the linux process and start a new one to bring up the new server Node?  This is an important question because TF has some per-process objects which will persist in the process between server instances, if you really try to restart in the same process.  The main example is memory allocators (see https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/process_state.h).  This could have strange effects.

Back to your requirements.  I'm guessing that you want Ignite to completely clean up old server instances and create new instances with no prior state.  If that's true, can Ignite not destroy the descendant process in which the old server is running (without waiting for the server to terminate) and create a new one in a new process?

If it's a high enough priority we should be able to make gRPC servers shut down cleanly, but until we try to do it, it's hard to say what kind of problems we might discover.

Comment (dmitrievanthony) [reply tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc]: Yes, I'm talking about linux process. I mentioned Apache Ignite just to describe the inspiration of the problem, don't think it has some specific requirements. The idea is to allow user to start a single java application (a java process) and do in this process something like:

```java
while (true) {
  Server server = new Server(...);
  server.start();
  // do something...
  server.stop();
}
```

without memory leaks and hangs. We have the current implementation of that via starting and killing separate Python processes (with `start()` server functionality written in Python) from Java. It works, but seems a bit to tricky compare with an ability to `start` and `stop` servers inside a single java process.

Comment (dmitrievanthony) [reply tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc]: Hi, @mrry, @poxvoculi. Do you have any comments regarding this issue?

Comment (mrry) [reply tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc]: The code in the PR doesn't solve the problem. It would need something like @poxvoculi's suggestion about a lame-duck mode, plus support for cancelling all in-flight requests (which isn't currently implemented). It would also need some refactoring to the device-management code, which uses per-process singletons at various points, so that all the required objects become owned by the server.

The process-based approach that @poxvoculi suggested would be more robust, at the expense of a little more overhead on starting the server, and the absence of a fast path between the parent and child processes. For the latter, it's not clear if that's a requirement.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ymodak: It has been 27 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ymodak: It has been 42 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer : You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer : You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 29 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer : You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 18 days with no activity and the `awaiting review` label has been applied.

**Comment by rthadur** [comment]
@dmitrievanthony wondering if this PR still valid , if yes could you please re base your branch.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by rthadur** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!
